stdnum/de/handelsregisternummer.py

```diff
@@ -0,0 +1,253 @@
+# handelsregisternummer.py - functions for handling German company registry id
+# coding: utf-8
+#
+# Copyright (C) 2015 Holvi Payment Services Oy
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301 USA
+
+""" Handelsregisternummer (German company register number).
+
+The number consists of the court where the company has registered, the type
+of register and the individual number
+
+The type of the register is either HRA or HRB where the letter "B" stands
+for HR section B, where limited liability companies and corporations are
+entered (GmbH's and AG's). There is also a section HRA for business
+partnerships (OHG's, KG's etc.). In other words: businesses in section
+HRB are limited liability companies, while businesses in HRA have personally
+liable partners.
+
+>>> businessid.validate('Achen HRA 11223')
+u'Achen HRA 11223'
+
+>>> businessid.validate('0 HRB 44123')
+u'Achen HRB 44123'
+
+>>> businessid.validate('Achen HRC 44123')
+Traceback (most recent call last):
+  ...
+stdnum.exceptions.InvalidFormat: The number has an invalid format.
+
+"""
+
+import re
+from stdnum.exceptions import *
+from stdnum.util import clean
+
+
+def validate(number):
+    """
+    Validate the format of a German company registry number.
+
+    First split the provided number with the HRA or HRB. Expect
+    the first part to be the name of the court and the second
+    part to be the registry number. Returns a cleaned representation
+    of the number with the court, HRA/HRB and the number.
+    """
+    parts = re.split('HR(A|B)', number, flags=re.IGNORECASE)
+    if len(parts) != 3:
+        raise InvalidFormat()
+    if parts[1].upper() not in ['A', 'B']:
+        raise InvalidFormat()
+    court = clean(parts[0], ':').strip()
+    if court.lower() not in (name.lower() for name in GERMAN_COURTS):
+        try:
+            ordinal = int(court)
+            if ordinal < len(GERMAN_COURTS) and ordinal >= 0:
+                court = GERMAN_COURTS[ordinal]
+            else:
+                raise InvalidFormat()
+        except ValueError:
+            raise InvalidFormat()
+    else:
+        index = [name.lower() for name in GERMAN_COURTS].index(court.lower())
+        court = GERMAN_COURTS[index]
```

Comment:
this can raise an error, I guess it will be picked up by the calling is_valid. Also I think some deaccenting at this point both in the source and the value passed would be good, but I believe these values are going to come from a preset list to the API, which in that case is OK.